drivers/nrf9160_gps/nrf9160_gps.c

```diff
@@ -0,0 +1,525 @@
+#define AT_XSYSTEMMODE_RESPONSE "%XSYSTEMMODE:"
+#define AT_XSYSTEMMODE_PROTO "AT%%XSYSTEMMODE=%d,%d,%d,%d"
+#define AT_XSYSTEMMODE_PARAMS_COUNT 4
+#define AT_XSYSTEMMODE_GPS_PARAM_INDEX 2
+#define AT_CFUN_REQUEST "AT+CFUN?"
+#define AT_CFUN_RESPONSE "+CFUN:"
+#define AT_CFUN_0 "AT+CFUN=0"
+#define AT_CFUN_1 "AT+CFUN=1"
+#define FUNCTIONAL_MODE_ENABLED 1
+
+struct gps_drv_data {
+	gps_trigger_handler_t trigger_handler;
+	struct gps_trigger trigger;
+	struct k_mutex trigger_mutex;
+
+	atomic_t gps_is_active;
+
+	int socket;
+
+	K_THREAD_STACK_MEMBER(thread_stack,
+			      CONFIG_NRF9160_GPS_THREAD_STACK_SIZE);
+	struct k_thread thread;
+	struct k_sem thread_run_sem;
+};
+
+static struct gps_data fresh_nmea;
+static struct gps_data fresh_pvt;
+
+static void copy_pvt(struct gps_pvt *dest, nrf_gnss_pvt_data_frame_t *src)
+{
+
+	dest->latitude = src->latitude;
+	dest->longitude = src->longitude;
+	dest->altitude = src->altitude;
+	dest->accuracy = src->accuracy;
+	dest->speed = src->speed;
+	dest->heading = src->heading;
+	dest->flags = src->flags;
+	dest->datetime.year = src->datetime.year;
+	dest->datetime.month = src->datetime.month;
+	dest->datetime.day = src->datetime.day;
+	dest->datetime.hour = src->datetime.hour;
+	dest->datetime.minute = src->datetime.minute;
+	dest->datetime.seconds = src->datetime.seconds;
+	dest->datetime.ms = src->datetime.ms;
+
+	for (size_t i = 0;
+	     i < MIN(NRF_GNSS_MAX_SATELLITES, GPS_MAX_SATELLITES); i++) {
+		dest->sv[i].sv = src->sv[i].sv;
+		dest->sv[i].cn0 = src->sv[i].cn0;
+		dest->sv[i].elevation = src->sv[i].elevation;
+		dest->sv[i].azimuth = src->sv[i].azimuth;
+		dest->sv[i].flags = src->sv[i].flags;
+		dest->sv[i].signal = src->sv[i].signal;
+	}
+}
+
+static bool is_fix(struct gps_pvt *pvt)
+{
+	return ((pvt->flags & NRF_GNSS_PVT_FLAG_FIX_VALID_BIT)
+		== NRF_GNSS_PVT_FLAG_FIX_VALID_BIT);
+}
+
+static u64_t fix_timestamp;
+
+static void print_satellite_stats(nrf_gnss_data_frame_t *pvt_data)
+{
+	u8_t  tracked          = 0;
+	u8_t  in_fix           = 0;
+	u8_t  unhealthy        = 0;
+
+	for (int i = 0; i < NRF_GNSS_MAX_SATELLITES; ++i) {
+
+		if ((pvt_data->pvt.sv[i].sv > 0) &&
+		    (pvt_data->pvt.sv[i].sv < 33)) {
+
+			tracked++;
+
+			if (pvt_data->pvt.sv[i].flags &
+					NRF_GNSS_PVT_FLAG_FIX_VALID_BIT) {
+				in_fix++;
+			}
+
+			if (pvt_data->pvt.sv[i].flags &
+					NRF_GNSS_SV_FLAG_UNHEALTHY) {
+				unhealthy++;
+			}
+		}
+	}
+
+	LOG_DBG("Tracking: %d Using: %d Unhealthy: %d", tracked,
+						       in_fix,
+						       unhealthy);
+
+	LOG_DBG("Seconds since last fix %lld",
+			(k_uptime_get() - fix_timestamp) / 1000);
+}
+
+static void gps_thread(int dev_ptr)
+{
+	struct device *dev = INT_TO_POINTER(dev_ptr);
+	struct gps_drv_data *drv_data = dev->driver_data;
+	int len;
+	nrf_gnss_data_frame_t raw_gps_data;
+	bool trigger_send = false;
+wait:
+	k_sem_take(&drv_data->thread_run_sem, K_FOREVER);
+
+	while (true) {
+		len = recv(drv_data->socket, &raw_gps_data,
```

Comment:
```suggestion
		len = nrf_recv(drv_data->socket, &raw_gps_data,
```

I've been having problems with the driver hanging/stopping at this **recv**. From what I can find on Devzone it seems like using **nrf_socket** and **net/socket** "mixed" can lead to undefined behavior as **net/socket** isn't fully compatible with the nRF9160 yet.

In this case the GPS socket is opened with **nrf_socket**, hence **nrf_recv** should be used later on. Did some testing with this and it seemed more stable.